File: hsg12-tinh/2024-2025-tayninh/cake/cake.cpp
#include <iostream>
#include <limits>
#include <vector>
#include <deque>
#include <algorithm>

#define input_file "cake.inp"
#define output_file "cake.out"

using namespace std;

typedef long long int lli;

// hằng số âm vô cực
const lli INF = numeric_limits<lli>::min();

int n, m, k;
vector<lli> A;

// điểm thưởng cao nhất cần tìm
lli max_score = 0;


void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    freopen(input_file, "r", stdin);

    cin >> n >> m >> k;

    A.resize(n + 1);

    for (int i = 1; i < n + 1; ++i)
    {
        cin >> A[i];
    }
}


void process()
{
    // Khởi tạo bảng quy hoạch động gồm k + 1 lượt chọn (hàng)
    // Mỗi hàng gồm n + 1 điểm thưởng cao nhất (cột)
    // D[choice][cake] là điểm thưởng cao nhất sau choice lượt chọn, kết thúc ở bánh thứ cake
    vector<vector<lli>> D(k + 1, vector<lli>(n + 1, INF));

    // Khởi tạo lượt chọn 1 (hàng 1)
    for (int cake = 1; cake < n + 1; ++cake)
    {
        D[1][cake] = A[cake];
    }

    // Điền bảng quy hoạch động từ lượt (hàng) 2 đến k
    // Duyệt từng lượt chọn (hàng)
    for (int choice = 2; choice < k + 1; ++choice)
    {
        // Khai báo hàng đợi hai đầu lưu chỉ số cake của lượt choice - 1
        deque<int> Q;

        // Duyệt từng bánh (cột)
        for (int cake = 1; cake < n + 1; ++cake)
        {
            if (D[choice - 1][cake - 1] != INF)
            {
                // Loại bỏ các D[][] nhỏ hơn ở cuối hàng đợi (tức bên phải)
                while (!Q.empty() && D[choice - 1][Q.back()] <= D[choice - 1][cake - 1])
                    Q.pop_back();

                // Nạp cake - 1 vào cuối hàng đợi
                Q.push_back(cake - 1);
            }

            // Loại bỏ chỉ số ở đầu hàng đợi Q vì nó nằm ngoài cửa sổ [cake - m, cake - 1]
            if (!Q.empty() && Q.front() < cake - m) Q.pop_front();

            // Tính D[choice][cake]
            if (!Q.empty())
            {
                if (D[choice - 1][Q.front()] != INF)
                    D[choice][cake] = D[choice - 1][Q.front()] + (lli)choice * A[cake];
            }
        }
    }

    // Duyệt các điểm cao nhất trong lượt k (hàng k) để lấy ra kết quả cao nhất
    max_score = *max_element(D[k].begin(), D[k].end());
}


void output()
{
    freopen(output_file, "w", stdout);

    cout << max_score;
}


int main()
{
    input();
    process();
    output();

    return 0;
}

File: hsg12-tinh/2024-2025-tayninh-buoi-1/cake/cake.py
from pathlib import Path
from collections import deque

dir = Path(__file__).parent
input_file = dir / 'cake.inp'
output_file = dir / 'cake.out'

# hằng số âm vô cực
INF = float('-inf')

n = m = k = 0
A = []

# điểm thưởng cao nhất cần tìm
max_score = 0


def input():
    global n, m , k, A

    with open(input_file, 'r') as f:
        n, m, k = map(int, f.readline().split())

        A = [0] + list(map(int, f.readline().split()))


def process():
    global n, m, k, A, max_score

    # Khởi tạo bảng quy hoạch động gồm k + 1 lượt chọn (hàng)
    # Mỗi hàng gồm n + 1 điểm thưởng cao nhất (cột)
    # D[choice][cake] là điểm thưởng cao nhất sau choice lượt chọn, kết thúc ở bánh thứ cake
    D = [[INF]  * (n + 1) for _ in range(k + 1)]

    # Khởi tạo lượt chọn 1 (hàng 1)
    for cake in range(1, n + 1):
        D[1][cake] = A[cake]

    # Điền bảng quy hoạch động từ lượt (hàng) 2 đến k
    # Duyệt từng lượt chọn (hàng)
    for choice in range(2, k + 1):
        # Khai báo hàng đợi hai đầu lưu chỉ số cake của lượt choice - 1
        Q = deque()

        # Duyệt từng bánh (cột)
        for cake in range(1, n + 1):
            if D[choice - 1][cake - 1] != INF:
                # Loại bỏ các D[][] nhỏ hơn ở cuối hàng đợi (tức bên phải)
                while Q and D[choice - 1][Q[-1]] <= D[choice - 1][cake - 1]:
                    Q.pop()

                # Nạp cake - 1 vào cuối hàng đợi
                Q.append(cake - 1)

            # Loại bỏ chỉ số ở đầu hàng đợi Q vì nó nằm ngoài cửa sổ [cake - m, cake - 1]
            if Q and Q[0] < cake - m:
                Q.popleft()

            # Tính D[choice][cake]
            if Q:
                if D[choice - 1][Q[0]] != INF:
                    D[choice][cake] = D[choice - 1][Q[0]] + choice * A[cake]

    # Duyệt các điểm cao nhất trong lượt k (hàng k) để lấy ra kết quả cao nhất
    max_score = max(D[k])


def output():
    global max_score

    with open(output_file, 'w') as f:
        f.write(f'{max_score}')


if __name__ == '__main__':
    input()
    process()
    output()
